### ISSUE TITLE
Ability to trace Tcl commands executed by Tkinter

### ISSUE THREAD

**Issue by serhiy-storchaka** [python/cpython#71592]
BPO | [27405](https://bugs.python.org/issue27405)
--- | :---
Nosy | @terryjreedy, @ned-deily, @serhiy-storchaka, @csabella
Files | <li>[tkinter_trace_tcl.patch](https://bugs.python.org/file43592/tkinter_trace_tcl.patch "Uploaded as text/plain at 2016-06-30.21:53:18 by @serhiy-storchaka")</li><li>[tkinter_trace_tcl2.patch](https://bugs.python.org/file44560/tkinter_trace_tcl2.patch "Uploaded as text/plain at 2016-09-11.17:41:19 by @serhiy-storchaka")</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = 'https://github.com/serhiy-storchaka'
closed_at = None
created_at = <Date 2016-06-28.08:06:00.853>
labels = ['3.7', 'type-feature', 'expert-tkinter']
title = 'Ability to trace Tcl commands executed by Tkinter'
updated_at = <Date 2018-10-17.22:53:22.982>
user = 'https://github.com/serhiy-storchaka'
```

bugs.python.org fields:
```python
activity = <Date 2018-10-17.22:53:22.982>
actor = 'cheryl.sabella'
assignee = 'serhiy.storchaka'
closed = False
closed_date = None
closer = None
components = ['Tkinter']
creation = <Date 2016-06-28.08:06:00.853>
creator = 'serhiy.storchaka'
dependencies = []
files = ['43592', '44560']
hgrepos = []
issue_num = 27405
keywords = ['patch']
message_count = 16.0
messages = ['269425', '269447', '269626', '269631', '269697', '269699', '269702', '269745', '275813', '275983', '275988', '275991', '275992', '276000', '276007', '327915']
nosy_count = 4.0
nosy_names = ['terry.reedy', 'ned.deily', 'serhiy.storchaka', 'cheryl.sabella']
pr_nums = []
priority = 'normal'
resolution = None
stage = 'test needed'
status = 'open'
superseder = None
type = 'enhancement'
url = 'https://bugs.python.org/issue27405'
versions = ['Python 3.7']
```

</p></details>

### Linked PRs
* gh-118291
* gh-118662
* gh-118664

**Comment by serhiy-storchaka** [comment]
It stuck for a long time because I was hesitant about the user interface. But every time I investigate a problem, I regret that this feature was not added. So I decided to add it as a private feature for our internal use. When we have some experience, we can add a public interface.

For the same reason, it is better to backport it.

**Comment by terryjreedy** [comment]
The merged interface: set tkinter._debug = True before creating the root window sets tracing (l.2454) with tkinter._print_command (l.5=2543), which prints every executed Tcl command .

Creating a branch and setting _debug at the top of pyshell.py and running `python -m idlelib` (in main repository) prints 100s of lines to Command Prompt for the creation of Shell:  lots of menu, bind, and handler calls.  It eventually settles into repeated Shell 50 ms socket polling loop calls, with 3 calls per poll:
```
proc 2199050013168poll_subprocess {} {<bound method CallWrapper.__call__ of <tkinter.CallWrapper object at 0x0000020001BDFCA0>>}
after 50 2199050013168poll_subprocess  # last action in poll_subprocess
rename 2199050013936poll_subprocess {}  # internal part of after call?
```
`poll_subprocess` defined about l.582; loop started l.486; 50 ms poll_interval set on l.950

I successfully stopped the loop with ctrl-C.  To use this with IDLE, to see for instance the calls resulting from a particular post-startup action, a custom `_print_command` would be needed that skips most of the editor window setup commands and all of the polling commands.  Looks quite doable.  The last non-poll call may have been "rename 2199051334160recolorize {}".  There might be a colorizer loop started later.

Sorry I did not review earlier.
